scanner/datastore/postgres/vuln_update.go

```diff
@@ -3,48 +3,81 @@ package postgres
 import (
 	"context"
 	"errors"
-	"fmt"
-	"net/http"
-	"strings"
 	"time"
 
 	"github.com/jackc/pgx/v4"
 )
 
-const lastVulnUpdateKey = `last-vuln-update`
-
-// GetLastVulnerabilityUpdate retrieves the last vulnerability update from the database.
+// GetLastVulnerabilityUpdate implements MatcherMetadataStore.GetLastVulnerabilityUpdate.
 //
-// The returned time will be in the form of http.TimeFormat.
+// Assumes the last update is the oldest update timestamp in the table.
 func (m *matcherMetadataStore) GetLastVulnerabilityUpdate(ctx context.Context) (time.Time, error) {
-	const selectTimestamp = `SELECT timestamp FROM last_vuln_update WHERE key = $1`
-
-	var t string
-	row := m.pool.QueryRow(ctx, selectTimestamp, lastVulnUpdateKey)
+	const selectTimestamp = `
+		SELECT update_timestamp
+		FROM last_vuln_update
+		WHERE update_timestamp IS NOT NULL ORDER BY update_timestamp LIMIT 1;`
+	row := m.pool.QueryRow(ctx, selectTimestamp)
+	var t time.Time
 	err := row.Scan(&t)
 	if errors.Is(err, pgx.ErrNoRows) {
 		return time.Time{}, nil
 	}
 	if err != nil {
 		return time.Time{}, err
 	}
+	return t.UTC(), nil
+}
 
-	timestamp, err := time.Parse(http.TimeFormat, strings.TrimSpace(t))
+// SetLastVulnerabilityUpdate implements MatcherMetadataStore.SetLastVulnerabilityUpdate.
+//
+// We use one row for each updater, keyed by their name.
+func (m *matcherMetadataStore) SetLastVulnerabilityUpdate(ctx context.Context, timestamp time.Time, updaterName string) error {
+	const insertTimestamp = `
+		INSERT INTO last_vuln_update (key, update_timestamp)
+		VALUES ($1, $2)
+		ON CONFLICT (key) DO UPDATE SET update_timestamp = $2;`
+	_, err := m.pool.Exec(ctx, insertTimestamp, updaterName, timestamp.UTC().Truncate(time.Second))
 	if err != nil {
-		return time.Time{}, fmt.Errorf("invalid timestamp: %w", err)
+		return err
 	}
 
-	return timestamp, nil
+	return nil
 }
 
-// SetLastVulnerabilityUpdate sets the last vulnerability update timestamp.
-func (m *matcherMetadataStore) SetLastVulnerabilityUpdate(ctx context.Context, timestamp time.Time) error {
-	const insertTimestamp = `INSERT INTO last_vuln_update (key, timestamp) VALUES ($1, $2) ON CONFLICT (key) DO UPDATE SET timestamp = $2`
+// GetOrInitLastVulnerabilityUpdate implements MatcherMetadataStore.GetOrInitLastVulnerabilityUpdate
+func (m *matcherMetadataStore) GetOrInitLastVulnerabilityUpdate(ctx context.Context, updaterName string) (time.Time, error) {
+	// SQL to try inserting a new row with zero value, in case the row does not exist.
+	const insertIfNotExist = `
+		INSERT INTO last_vuln_update (key, update_timestamp)
+		VALUES ($1, $2)
+		ON CONFLICT (key) DO NOTHING;`
+	_, err := m.pool.Exec(ctx, insertIfNotExist, updaterName, time.Time{})
```

Comment:
Should this be something like `sql.NullTime`? Maybe they're functionality equivalent in postgres - I'm not sure. I'm worried `time.Time{}` will be interpreted as "the oldest update" regarding `GetLastVulnerabilityUpdate`

Reply:
>  Maybe they're functionality equivalent in postgres
If they're not functionality equivalent, but your changes still work, and you want to keep it this way, my next question would be: do we need the timestamp to be nullable?

Reply:
> Should this be something like sql.NullTime? Maybe they're functionality equivalent in postgres - I'm not sure. I'm worried time.Time{} will be interpreted as "the oldest update" regarding GetLastVulnerabilityUpdate

Thanks for noticing that. Actually, this should be set to "the oldest update." Initially, yes, 0, but when introducing new updaters, we should always initialize to the latest update. I will update this.

Reply:
> do we need the timestamp to be nullable?

Because we might migrate this to an existing DB, and I don't want to step in the old row with the global last update timestamp.

Reply:
I was under the assumption that the old matcher datastore set/update would fail if the column did not have a default. But it seems that's not true, so I set the default to `to_timestamp(0)`, and converted any previous timestamp to the new one.